### ISSUE TITLE
Possible enhancement for `[_]` in `Data.List/Vec/Vector`

### ISSUE THREAD

**Issue by MatthewDaggitt** [agda/agda-stdlib#1497]
There has been some informal discussion about changing the singleton list/vector `[_]` to a pattern synonym so that you can use it on the left hand side and write things like:
```agda
intersperse : List A -> A -> List A
intersperse [] _ = []
intersperse [ x ] _ = [ x ]
intersperse (x :: xs) y = x :: y :: intersperse xs y
```
However, swapping functions for (untyped) pattern synonyms is not backwards compatible so it hasn't been carried out yet.

In the meantime I propose that we redefine `[_]` to make it more powerful so we can write an arbitrary number of objects inside of the brackets as in other programming languages, e.g. `[ 5 , 2 , 3 ]`. This can be achieved by using `_^_` from `Data.Vec.Recursive` as follows:
```
[_] : A ^ (suc n) → Vec A (suc n)
[_] x = toVec _ x
```
This _is_ a backwards compatible change, and I don't think precludes making it a pattern in v2.0? @gallais 

(although it may need a bit of rejigging to fix the dependency cycles)

**Comment by gallais** [comment]
Oooh that's sneaky! :heart: 

I was expecting some sort of horrible overloading of `[_` as the identity function,
`_,_` as `_::_` and `_]` as nil. This is much better.

I do think it should be fine. Functions are never considered on the LHS so it should
not clash there (it will however clash on the RHS where both the pattern and the `[_]`
function will be valid parses)

**Comment by jamesmckinna** [comment]
What is the current status of `[_]` as a pattern synonym? 
is there a separate issue for it, and/or is there any likelihood of being able to add such a synonym 'without tears'?

**Comment by MatthewDaggitt** [comment]
I believe it's blocked by the introduction of typed synonyms for which PRs have been lurking in the Agda repo for many years but have never landed.

**Comment by jamesmckinna** [comment]
See also: https://github.com/agda/agda/issues/7061

**Comment by jamesmckinna** [comment]
And now #3095 which ~~would definitely clash with~~ solves the 'singleton pattern' part of this proposal, but would not admit the clever `Recursive` trick... but seeing the fate of #1499 perhaps that's OK?
